### PR TITLE
[FLINK-31897] Fix the unstable test ClientTest#testRequestUnavailableHost

### PR DESCRIPTION
**Pull request by TanYuxin-tyx**
## What is the purpose of the change

*Fix the unstable test ClientTest#testRequestUnavailableHost.*

## Brief change log

  - *Migrate ClientTest to JUnit5*
  - *Fix the unstable test ClientTest#testRequestUnavailableHost*

## Verifying this change

This change is a trivial test fix without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yno)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java [hunk 1]
```diff
@@ -248,12 +251,30 @@ void testRequestUnavailableHost() throws Exception {
...
+                    break;
+                } catch (Exception e) {
+                    if (ExceptionUtils.findThrowable(e, BindException.class).isPresent()
+                            && numRetry <= numTryCreateSocketAddress) {
+                        numRetry++;
+                    } else {
+                        throw e;
+                    }
+                }
+            }
```
Comment (reswqa) [inline flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java:275]: Actually, flink has NetUtils to find a free port 

`int port = NetUtils.getAvailablePort().getPort();`

**Review by reswqa — COMMENTED** [review]
Thanks @TanYuxin-tyx for the fix, I left some comments, PTAL.

File: flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java [hunk 2]
```diff
@@ -721,29 +740,34 @@ public void testClientServerIntegration() throws Throwable {
...
-                    assertThat(t, exceptionMatcher);
+                } catch (Throwable throwable) {
+                    assertThat(
+                                    ExceptionUtils.findThrowable(
+                                                            throwable, ClosedChannelException.class)
+                                                    .isPresent()
+                                            || ExceptionUtils.findThrowable(
+                                                            throwable, IllegalStateException.class)
+                                                    .isPresent())
+                            .isTrue();
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java]: I have used `assertThat` instead.

File: flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java [hunk 3]
```diff
@@ -506,49 +525,49 @@ public void testServerClosesChannel() throws Exception {
...
+                assertThatFuture(future).eventuallyFails();
                 future.get();
-                fail("Did not throw expected server failure");
             } catch (ExecutionException e) {
-                if (!(e.getCause() instanceof ClosedChannelException)) {
-                    fail("Did not throw expected Exception");
-                }
+                assertThat(e.getCause())
+                        .withFailMessage("Did not throw expected Exception")
+                        .isInstanceOf(ClosedChannelException.class);
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java]: Use `FlinkAssertions.assertThatFuture(future)` instead.

File: flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java [hunk 4]
```diff
@@ -257,33 +251,56 @@ public void testRequestUnavailableHost() throws Exception {
...
+            assertThat(future).isNotNull();
+            future.whenComplete(
+                    (result, exception) -> {
+                        assertThat(result).isNull();
+                        assertThat(
+                                        ExceptionUtils.findThrowable(
+                                                        exception, ConnectException.class)
+                                                .isPresent())
+                                .isTrue();
+                    });
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java]: Use `assertThatThrownBy(future::get).hasRootCauseInstanceOf(ConnectException.class)` instead.

File: flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java (hunk 1 above)
Comment (TanYuxin-tyx) [reply flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java]: Good point. Fixed.

**Comment by TanYuxin-tyx** [comment]
Thanks for helping review, @reswqa I have addressed the comments. 
Please take a look again, thanks.

File: flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java [hunk 5]
```diff
@@ -721,29 +690,29 @@ public void testClientServerIntegration() throws Throwable {
...
             for (Future<Void> future : taskFutures) {
                 try {
                     future.get();
-                    fail("Did not throw expected Exception after shut down");
-                } catch (ExecutionException t) {
-                    assertThat(t, exceptionMatcher);
+                } catch (Throwable throwable) {
+                    assertThat(throwable.getCause().getCause())
+                            .isInstanceOfAny(
+                                    ClosedChannelException.class, IllegalStateException.class);
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java]: I think it makes sense. 

Fixed it by a **fixup** commit.

**Review by reswqa — APPROVED** [review]
LGTM % CI.

**Comment by TanYuxin-tyx** [comment]
@reswqa Thanks for the quick review and the quick reply.

File: flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java [hunk 6]
```diff
@@ -257,33 +244,33 @@ public void testRequestUnavailableHost() throws Exception {
         try {
             client = new Client<>("Test Client", 1, serializer, stats);
 
-            InetSocketAddress serverAddress = new InetSocketAddress(InetAddress.getLocalHost(), 0);
+            InetSocketAddress serverAddress =
+                    new InetSocketAddress(
+                            InetAddress.getLocalHost(), NetUtils.getAvailablePort().getPort());
```
Comment (zentol) [inline flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java:249]: What makes you believe that this is more stable?

**Comment by zentol** [comment]
Is this even about stability? Or just that the reporter has a restrictive setup where it's not allowed to bind to port 0 or something.

**Comment by TanYuxin-tyx** [comment]
@zentol Thanks for joining the review. 

`InetSocketAddress` only uses a random port when the **`bind`** operation occurs, as seen in the InetSocketAddress docs. 

But in this test, it does not trigger the `bind` operation obviously, so it uses port 0 directly and causes the test to be unstable when port 0 is occupied. I tested locally and the test often fails because of the `BindException`.

If you do not agree on the fix, maybe another fixing method is that use the `bind` to trigger to choose an available port like this to replace the line `InetSocketAddress address = new InetSocketAddress(InetAddress.getLocalHost(), 0)`.
```
            ServerSocket serverSocket = new ServerSocket();
            InetSocketAddress address = new InetSocketAddress(InetAddress.getLocalHost(), 0);
            System.out.println("The port 1 " + address.getPort());
            serverSocket.bind(address);
            serverSocket.close();

            System.out.println("The port 2 " + serverSocket.getLocalPort());
            InetSocketAddress serverAddress =
                    new InetSocketAddress(InetAddress.getLocalHost(), serverSocket.getLocalPort());
```
(These `System.out.println` should be removed. They are only used to see the results.)

You can see that `port 1` is 0 and `port 2` is another available port.  Then the test will be stable.

WDYT about the new fixing method?

**Comment by zentol** [comment]
> maybe another fixing method is that use the bind to trigger to choose an available port

I don't think this'd make the test stable.

The test requires a port against which it can send a request, and we _know_ it won't get a response.
For that we need to fully control that port, because otherwise some stray process may happen to bind to that port and answer the request.

What if we'd create UDP socket on a random port and use that instead?

```
            DatagramSocket s = new DatagramSocket();

            InetSocketAddress serverAddress =
                    new InetSocketAddress(InetAddress.getLocalHost(), s.getLocalPort());
```

**Comment by TanYuxin-tyx** [comment]
@zentol 
Apologies for my lack of understanding, but I am unclear on how the change ensures that the port is available.

When creating a `DatagramSocket s = new DatagramSocket()`, a UDP socket is generated. However, when an `InetSocketAddress` is created, it only specifies an **Integer** type of the port. How does the `InetSocketAddress` differentiate between UDP and TCP?

In my opinion, if we do not change the client's initialization process, `client.sendRequest` will still utilize TCP. Therefore, I fail to see how using `new DatagramSocket()` versus `new ServerSocket` makes a difference. 

Could you help take a look again?  Thanks a lot.

**Comment by TanYuxin-tyx** [comment]
I agree that using port 0 is usually the best approach to resolve conflicts, but it indeed does not work well in this test.

> The UDP socket only ensures no other process binds to that port in the mean time.

I don't think so. Because TCP and UDP can use the same port at the same time. For example, if the UDP server has used port 23456, the TCP can also use the same port at the same time.

**Comment by zentol** [comment]
> if the UDP server has used port 23456, the TCP can also use the same port at the same time.

Ah that's a very good point.

Maybe we should not bother with ports at all but rather focus on hostnames. The set of resolvable hostnames should change a lot less than ports; a bogus hostname (that we can check for beforehand whether it resolves to anything) could be a better approach.
This hostname could be awfully specific like "flink-qs-client-test-unavailable-host" so the chances of conflicts are basically 0.

**Comment by TanYuxin-tyx** [comment]
@zentol Thanks for the quick replies and ideas. 

Since no real servers are created based on the server address, I think the given fixed port is enough. And the hostname has been updated. 

Could you please help take a look again? Thanks.

**Comment by TanYuxin-tyx** [comment]
Thanks @zentol and @reswqa  for the quick reviews.

**Pull request merged**